File: src/engine/graph/eval.ts
/**
 * eval.ts
 *
 * Evaluates the whole document. It sorts every slot into dependency order and
 * then evaluates each one in turn, so a slot always reads inputs that have
 * already been recomputed.
 * The depth first sort uses an explicit stack, so long dependency chains
 * retain their traversal order without exhausting the JavaScript call stack.
 *
 * The three kinds of slot go through this single pass. A literal returns the
 * value it stores, a formula evaluates its AST, and a derived slot calls the
 * compute function its schema declares. Because all three take the same path,
 * a derived value is never one evaluation behind the literal that feeds it.
 *
 * Evaluation never throws. A formula that fails leaves an error value in its
 * own slot and the pass carries on, so one broken formula cannot blank the
 * rest of the document.
 *
 * Nothing here knows about any object type. A script node is one more derived
 * slot as far as this file is concerned, so a new primitive can arrive without
 * any change to the evaluator.
 *
 * Engine-layer code: pure logic with no DOM, window or canvas access, so the
 * tests run headless and the file can move to Rust later.
 */

import type { Address } from "../address.ts";
import { NULL_EVAL_CONTEXT, type EvalContext } from "../eval-context.ts";
import type { FormulaAst } from "../formula/ast.ts";
import { evaluate as evaluateFormulaAst, type ReadRange, type ReadSlot } from "../formula/eval.ts";
import { enumerateRangeCellAddresses, isInExtentTableCellAddress, isRangeEnumerationError } from "../primitives/table.ts";
import { getObjectSchema, resolveDerivedSlots, type DerivedSlotSchema } from "../primitives/schema.ts";
import { addressKey, type Edge } from "./edge.ts";
import { slotKey, type GraphObject, type Slot, type Value } from "./node.ts";

/**
 * Evaluates every slot of every object, in topological order. It returns the
 * new object list. An error becomes an error value in a slot. It never
 * throws, because one bad formula does not stop the rest.
 */
export function evaluate(
  objects: readonly GraphObject[],
  edges: readonly Edge[],
  context: EvalContext = NULL_EVAL_CONTEXT,
): readonly GraphObject[] {
  const nodesByKey = new Map<string, { readonly object: GraphObject; readonly key: string }>();
  for (const object of objects) {
    for (const key of Object.keys(object.slots)) {
      nodesByKey.set(`${object.id}::${key}`, { object, key });
    }
  }

  const outgoing = new Map<string, string[]>();
  for (const edge of edges) {
    const sourceKey = addressKey(edge.sourceSlot);
    const arcs = outgoing.get(sourceKey);
    if (arcs === undefined) {
      outgoing.set(sourceKey, [addressKey(edge.dependentSlot)]);
    } else {
      arcs.push(addressKey(edge.dependentSlot));
    }
  }

  const visited = new Set<string>();
  const postorder: string[] = [];
  const stack: { key: string; next: number }[] = [];
  for (const nodeKey of nodesByKey.keys()) {
    if (visited.has(nodeKey)) continue;
    visited.add(nodeKey);
    stack.push({ key: nodeKey, next: 0 });
    while (stack.length > 0) {
      const frame = stack[stack.length - 1]!;
      const next = outgoing.get(frame.key)?.[frame.next++];
      if (next === undefined) {
        postorder.push(frame.key);
        stack.pop();
      } else if (!visited.has(next)) {
        visited.add(next);
        stack.push({ key: next, next: 0 });
      }
    }
  }
  const topologicalOrder = postorder.slice().reverse();

  const evaluatedValues = new Map<string, Value>();

  const updatedSlotsByObjectId = new Map<string, Record<string, Slot>>();
  function slotsFor(objectId: string): Record<string, Slot> {
    const existing = updatedSlotsByObjectId.get(objectId);
    if (existing !== undefined) {
      return existing;
    }
    const created: Record<string, Slot> = {};
    updatedSlotsByObjectId.set(objectId, created);
    return created;
  }

  for (const nodeKey of topologicalOrder) {
    const node = nodesByKey.get(nodeKey);
    if (node === undefined) {
      continue;
    }
    const { object, key } = node;
    const slot = object.slots[key];
    if (slot === undefined) {
      continue;
    }

    const value = evaluateSlot(object, key, slot, objects, edges, evaluatedValues, context);
    evaluatedValues.set(nodeKey, value);
    slotsFor(object.id)[key] = nextSlot(slot, value);
  }

  return objects.map((object) => ({
    ...object,
    slots: { ...object.slots, ...slotsFor(object.id) },
  }));
}

function evaluateSlot(
  object: GraphObject,
  key: string,
  slot: Slot,
  objects: readonly GraphObject[],
  edges: readonly Edge[],
  evaluatedValues: ReadonlyMap<string, Value>,
  context: EvalContext,
): Value {
  switch (slot.kind) {
    case "literal":
      return slot.value;
    case "formula":
      return evaluateFormula(slot.ast, objects, evaluatedValues);
    case "derived":
      return evaluateDerivedSlot(object, key, objects, edges, evaluatedValues, context);
  }
}

function nextSlot(slot: Slot, value: Value): Slot {
  switch (slot.kind) {
    case "literal":
      return slot;
    case "formula":
      return { kind: "formula", ast: slot.ast, value };
    case "derived":
      return { kind: "derived", value };
  }
}

function evaluateFormula(ast: FormulaAst, objects: readonly GraphObject[], evaluatedValues: ReadonlyMap<string, Value>): Value {
  const read: ReadSlot = (address) => {
    const value = evaluatedValues.get(addressKey(address));
    return isEmptyInExtentCell(address, value, objects) ? 0 : value;
  };
  return evaluateFormulaAst(ast, read, buildRangeReader(objects, evaluatedValues));
}

function isEmptyInExtentCell(address: Address, rawValue: Value | undefined, objects: readonly GraphObject[]): boolean {
  return (rawValue === undefined || rawValue === null) && isInExtentTableCellAddress(address, objects);
}

function buildRangeReader(objects: readonly GraphObject[], evaluatedValues: ReadonlyMap<string, Value>): ReadRange {
  return (start, end) => {
    const tableObject = objects.find((candidate) => candidate.id === start.objectId);
    if (tableObject === undefined) {
      return { error: "#REF", message: "a range references an object that does not exist" };
    }
    const cellAddresses = enumerateRangeCellAddresses(start, end, tableObject);
    if (isRangeEnumerationError(cellAddresses)) {
      return cellAddresses;
    }
    const values: Value[] = [];
    for (const cellAddress of cellAddresses) {
      const value = evaluatedValues.get(addressKey(cellAddress));
      if (value === undefined || value === null) {
        continue;
      }
      values.push(value);
    }
    return values;
  };
}

function evaluateDerivedSlot(
  object: GraphObject,
  key: string,
  objects: readonly GraphObject[],
  edges: readonly Edge[],
  evaluatedValues: ReadonlyMap<string, Value>,
  context: EvalContext,
): Value {
  const schema = getObjectSchema(object.type);
  const schemaEntry = findDerivedSlotSchemaByKey(schema === undefined ? undefined : resolveDerivedSlots(object, schema.derivedSlots), key);
  if (schemaEntry === undefined) {
    return {
      error: "#REF",
      message: `no schema entry declares a derived slot "${key}" on type "${object.type}"`,
    };
  }

  const ownKey = addressKey({ objectId: object.id, path: schemaEntry.path });
  const declaredDependencyKeys = new Set(
    edges.filter((edge) => addressKey(edge.dependentSlot) === ownKey).map((edge) => addressKey(edge.sourceSlot)),
  );

  const read = (address: Address): Value | undefined => {
    const addressAsKey = addressKey(address);
    const rawValue = evaluatedValues.get(addressAsKey);
    if (isEmptyInExtentCell(address, rawValue, objects)) {
      return 0;
    }
    if (!declaredDependencyKeys.has(addressAsKey)) {
      return {
        error: "#REF",
        message: "derived slot's compute function read an address outside its declared dependencies",
      };
    }
    return rawValue;
  };

  const readRange = buildRangeReader(objects, evaluatedValues);

  return schemaEntry.compute(object, read, context, { readRange, objects });
}

function findDerivedSlotSchemaByKey(
  derivedSlots: readonly DerivedSlotSchema[] | undefined,
  key: string,
): DerivedSlotSchema | undefined {
  return derivedSlots?.find((entry) => slotKey(entry.path) === key);
}

File: crates/beheader-engine/src/graph/eval.rs
//! Evaluates the whole document. It sorts every slot into dependency order and
//! then evaluates each one in turn, so a slot always reads inputs that have
//! already been recomputed. The depth first sort uses an explicit stack, so long
//! dependency chains keep their traversal order without exhausting a call stack.
//!
//! The three kinds of slot go through this single pass. A literal answers with
//! the value it stores, a formula evaluates its tree, and a derived slot calls
//! the compute function its schema declares. Because all three take the same
//! path, a derived value is never one evaluation behind the literal that feeds
//! it.
//!
//! Evaluation never fails. A formula that cannot be computed leaves an error
//! value in its own slot and the pass carries on, so one broken formula cannot
//! blank the rest of the document.
//!
//! Nothing here knows about any object type. A script node is one more derived
//! slot as far as this file is concerned, so a new primitive can arrive with no
//! change to the evaluator.
//!
//! The pass reads the slot set and never writes it, which is rule 4. It answers
//! with the objects it was given, each slot carrying its new value and the same
//! key it arrived under.
//!
//! The port of `src/engine/graph/eval.ts`.

use std::collections::{HashMap, HashSet};

use crate::address::Address;
use crate::formula::ast::FormulaAst;
use crate::formula::eval::evaluate as evaluate_formula_ast;
use crate::graph::{Edge, address_key};
use crate::measure::Measurer;
use crate::model::{ErrorCode, ErrorValue, GraphObject, Slot, Value, slot_key};
use crate::primitives::schema::{
    DerivedSlotSchema, SlotComputeInputs, get_object_schema, resolve_derived_slots,
};
use crate::primitives::table::{enumerate_range_cell_addresses, is_in_extent_table_cell_address};

/// One slot of one object, named the way a traversal names it.
struct Node {
    object_index: usize,
    key: String,
}

/// Evaluates every slot of every object, in topological order. It answers with
/// the new object list. A failure becomes an error value in a slot, and the pass
/// carries on, because one bad formula does not stop the rest.
pub fn evaluate(
    objects: &[GraphObject<FormulaAst>],
    edges: &[Edge],
    measurer: Option<&dyn Measurer>,
) -> Vec<GraphObject<FormulaAst>> {
    // The nodes are kept in the order the objects carry their slots, because
    // the order a depth first sort starts from decides the order two slots that
    // do not depend on one another are evaluated in.
    let mut node_order: Vec<String> = Vec::new();
    let mut nodes: HashMap<String, Node> = HashMap::new();
    for (object_index, object) in objects.iter().enumerate() {
        for (key, _) in object.slots.iter() {
            let node_key = format!("{}::{}", object.id, key);
            if let std::collections::hash_map::Entry::Vacant(slot) = nodes.entry(node_key.clone()) {
                node_order.push(node_key);
                slot.insert(Node {
                    object_index,
                    key: key.to_string(),
                });
            }
        }
    }

    let mut outgoing: HashMap<String, Vec<String>> = HashMap::new();
    for edge in edges {
        outgoing
            .entry(address_key(&edge.source_slot))
            .or_default()
            .push(address_key(&edge.dependent_slot));
    }

    let mut visited: HashSet<String> = HashSet::new();
    let mut postorder: Vec<String> = Vec::new();
    let mut stack: Vec<(String, usize)> = Vec::new();
    for node_key in &node_order {
        if visited.contains(node_key) {
            continue;
        }
        visited.insert(node_key.clone());
        stack.push((node_key.clone(), 0));
        while let Some((key, next)) = stack.last_mut() {
            let taken = *next;
            *next += 1;
            let key = key.clone();
            let neighbour = outgoing.get(&key).and_then(|arcs| arcs.get(taken)).cloned();
            match neighbour {
                None => {
                    postorder.push(key);
                    stack.pop();
                }
                Some(neighbour) => {
                    if visited.insert(neighbour.clone()) {
                        stack.push((neighbour, 0));
                    }
                }
            }
        }
    }
    postorder.reverse();

    let mut evaluated: HashMap<String, Value> = HashMap::new();
    let mut updated: HashMap<String, Vec<(String, Slot<FormulaAst>)>> = HashMap::new();

    for node_key in &postorder {
        let Some(node) = nodes.get(node_key) else {
            continue;
        };
        let object = &objects[node.object_index];
        let Some(slot) = object.slots.get(&node.key) else {
            continue;
        };
        let value = evaluate_slot(
            object, &node.key, slot, objects, edges, &evaluated, measurer,
        );
        evaluated.insert(node_key.clone(), value.clone());
        updated
            .entry(object.id.clone())
            .or_default()
            .push((node.key.clone(), next_slot(slot, value)));
    }

    objects
        .iter()
        .map(|object| {
            let mut slots = object.slots.clone();
            if let Some(written) = updated.get(&object.id) {
                for (key, slot) in written {
                    slots.insert(key.as_str(), slot.clone());
                }
            }
            GraphObject {
                slots,
                ..object.clone()
            }
        })
        .collect()
}

fn evaluate_slot(
    object: &GraphObject<FormulaAst>,
    key: &str,
    slot: &Slot<FormulaAst>,
    objects: &[GraphObject<FormulaAst>],
    edges: &[Edge],
    evaluated: &HashMap<String, Value>,
    measurer: Option<&dyn Measurer>,
) -> Value {
    match slot {
        Slot::Literal { value } => value.clone(),
        Slot::Formula { ast, .. } => evaluate_formula(ast, objects, evaluated),
        Slot::Derived { .. } => {
            evaluate_derived_slot(object, key, objects, edges, evaluated, measurer)
        }
    }
}

/// The slot a pass leaves behind. A literal keeps what it stores, and the other
/// two carry the value the pass computed.
fn next_slot(slot: &Slot<FormulaAst>, value: Value) -> Slot<FormulaAst> {
    match slot {
        Slot::Literal { .. } => slot.clone(),
        Slot::Formula { ast, .. } => Slot::Formula {
            ast: ast.clone(),
            value,
        },
        Slot::Derived { .. } => Slot::Derived { value },
    }
}

/// True for a cell inside the size of a real table that holds nothing. Such a
/// cell reads as zero rather than as a missing slot, because an empty cell is
/// ordinary state and a sum across a sparse row would otherwise refuse.
fn is_empty_in_extent_cell(
    address: &Address,
    held: Option<&Value>,
    objects: &[GraphObject<FormulaAst>],
) -> bool {
    matches!(held, None | Some(Value::Null)) && is_in_extent_table_cell_address(address, objects)
}

fn evaluate_formula(
    ast: &FormulaAst,
    objects: &[GraphObject<FormulaAst>],
    evaluated: &HashMap<String, Value>,
) -> Value {
    let read = |address: &Address| {
        let held = evaluated.get(&address_key(address));
        if is_empty_in_extent_cell(address, held, objects) {
            return Some(Value::Number(0.0));
        }
        held.cloned()
    };
    let read_range = range_reader(objects, evaluated);
    evaluate_formula_ast(ast, &read, Some(&read_range))
}

/// The cells of a range, as the pass has them so far.
fn range_reader<'a>(
    objects: &'a [GraphObject<FormulaAst>],
    evaluated: &'a HashMap<String, Value>,
) -> impl Fn(&Address, &Address) -> Result<Vec<Value>, ErrorValue> + 'a {
    move |start, end| {
        let Some(table) = objects
            .iter()
            .find(|candidate| candidate.id == start.object_id)
        else {
            return Err(ErrorValue {
                error: ErrorCode::Ref,
                message: "a range references an object that does not exist".to_string(),
            });
        };
        let cells =
            enumerate_range_cell_addresses(start, end, table).map_err(|failure| ErrorValue {
                error: ErrorCode::Ref,
                message: failure.message,
            })?;
        let mut values = Vec::new();
        for cell in cells {
            match evaluated.get(&address_key(&cell)) {
                None | Some(Value::Null) => continue,
                Some(value) => values.push(value.clone()),
            }
        }
        Ok(values)
    }
}

fn evaluate_derived_slot(
    object: &GraphObject<FormulaAst>,
    key: &str,
    objects: &[GraphObject<FormulaAst>],
    edges: &[Edge],
    evaluated: &HashMap<String, Value>,
    measurer: Option<&dyn Measurer>,
) -> Value {
    let declared = get_object_schema::<FormulaAst>(object.object_type)
        .map(|schema| resolve_derived_slots(object, schema.derived_slots));
    let found = declared.and_then(|slots| find_derived_slot_by_key(slots, key));
    let Some(entry) = found else {
        return Value::Error(ErrorValue {
            error: ErrorCode::Ref,
            message: format!(
                "no schema entry declares a derived slot \"{key}\" on type \"{}\"",
                object.object_type.as_str()
            ),
        });
    };

    // A compute function reads only what its schema declared, so a dependency
    // the schema left out cannot reach it by accident and leave the graph
    // carrying an edge nobody derived.
    let own_key = address_key(&Address {
        object_id: object.id.clone(),
        path: entry.path.clone(),
    });
    let declared_keys: HashSet<String> = edges
        .iter()
        .filter(|edge| address_key(&edge.dependent_slot) == own_key)
        .map(|edge| address_key(&edge.source_slot))
        .collect();

    let read = |address: &Address| {
        let as_key = address_key(address);
        let held = evaluated.get(&as_key);
        if is_empty_in_extent_cell(address, held, objects) {
            return Some(Value::Number(0.0));
        }
        if !declared_keys.contains(&as_key) {
            return Some(Value::Error(ErrorValue {
                error: ErrorCode::Ref,
                message:
                    "derived slot's compute function read an address outside its declared dependencies"
                        .to_string(),
            }));
        }
        held.cloned()
    };
    let read_range = range_reader(objects, evaluated);
    let inputs = SlotComputeInputs {
        read: &read,
        read_range: Some(&read_range),
        objects,
        measurer,
    };
    (entry.compute)(object, &inputs)
}

fn find_derived_slot_by_key(
    slots: Vec<DerivedSlotSchema<FormulaAst>>,
    key: &str,
) -> Option<DerivedSlotSchema<FormulaAst>> {
    slots.into_iter().find(|entry| slot_key(&entry.path) == key)
}
